crate_universe/src/main.rs

```diff
@@ -36,49 +38,55 @@ fn main() -> anyhow::Result<()> {
-    let repo_name = opt.repo_name;
-    let output_path = opt.output_path;
-
-    trace!("Preprocessing config");
-    let mut resolver = config.preprocess()?;
 
+    let lockfile = &opt.lockfile;
     if opt.update_lockfile {
         if lockfile.is_none() {
-            eprintln!("Not updating lockfile for `crate_universe` repository with name \"{}\" because it has no `lockfile` attribute.", repo_name);
+            eprintln!("Not updating lockfile for `crate_universe` repository with name \"{}\" because it has no `lockfile` attribute.", opt.repo_name);
         }
     } else if let Some(lockfile) = lockfile {
-        let mut lockfile_format_version_line = String::new();
-        let mut lockfile_hash_line = String::new();
-        {
-            std::fs::File::open(&lockfile)
-                .map(BufReader::new)
-                .and_then(|mut f| {
-                    f.read_line(&mut lockfile_format_version_line)?;
-                    f.read_line(&mut lockfile_hash_line)?;
-                    Ok(())
-                })
-                .with_context(|| format!("Failed to read lockfile header from {:?}", lockfile))?;
-        }
-        if lockfile_format_version_line != "# rules_rust crate_universe file format 1\n" {
-            return Err(anyhow!("Unrecognized lockfile format"));
-        }
-        if let Some(lockfile_hash) = lockfile_hash_line.strip_prefix("# config hash ") {
-            if resolver.digest()? == lockfile_hash.trim() {
-                std::fs::copy(&lockfile, &output_path).with_context(|| {
-                    format!(
-                        "Failed to copy lockfile from {:?} to {:?}",
-                        lockfile, output_path
-                    )
-                })?;
-                return Ok(());
-            } else {
-                return Err(anyhow!("rules_rust_external: Lockfile at {} is out of date, please either:\n1. Re-run bazel with the environment variable `RULES_RUST_UPDATE_CRATE_UNIVERSE_LOCKFILE=true`, to update the lockfile\n2. Remove the `lockfile` attribute from the `crate_universe` repository rule with name \"{}\" to use floating dependency versions", lockfile.display(), repo_name));
-            }
-        } else {
-            return Err(anyhow!("Invalid lockfile"));
-        }
+        return reuse_lockfile(config, &lockfile, &opt);
+    }
+
+    generate_dependencies(config, &opt)
+}
+
+fn reuse_lockfile(config: Config, lockfile: &Path, opt: &Opt) -> anyhow::Result<()> {
+    trace!("Preprocessing config");
+    let mut resolver = config.preprocess()?;
+
+    let renderer = Renderer::new_from_lockfile(lockfile)?;
+
+    // TODO: Add lockfile versioning and check that here
+
+    if !renderer.matches_digest(&resolver.digest()?) {
+        return Err(anyhow!(indoc! { r#"
+            "rules_rust_external: Lockfile at {} is out of date, please either:
+            1. Re-run bazel with the environment variable `RULES_RUST_UPDATE_CRATE_UNIVERSE_LOCKFILE=true`, to update the lockfile
```

Comment:
```suggestion
            1. Re-run bazel with the environment variable `RULES_RUST_REPIN=true`, to update the lockfile
```

Again, not your fault, but turns out we changed this env var name but not the warning message!